### DISCUSSION TITLE
socket io and manually updating the cache

### DISCUSSION THREAD

**Discussion by hgulersen** [urql-graphql/urql#2075] [Q&A] [answered]
I am using a graphql backend where subscriptions are not an option (strapi cms). For real time functionality, I'm using Socket IO to emit events to the client from the server. Currently I have the following structure: 

1. graphql query to fetch the data - let's say result is `[1, 2, 3]` 
2. socket io notifies that there's new data to fetch 
3. I use urql's `reexecuteQuery` with `{ requestPolicy: "cache-and-network" }`, the new result becomes `[1, 2, 3, 4]` 

In practice, this works, but it means that I have to re-fetch all the data (`[1, 2, 3]`) that I already have in cache just for the sake of getting the last piece of data (`4`), which could have been transmitted by socket io in the first place. My question is, since I have `[1, 2, 3]` already in cache thanks to my first query, if I emit `4` from socket io, can I update the local cache and append the new data somehow? It would save me an additional roundtrip + the cost of re-fetching the initial data.

**Comment by mwangaben** [comment]
Please assist on this since i would love to do the same thing on my app

**Comment by JoviDeCroock** [comment, marked as answer]
Hey,

What kind off cache are you using? Default or normalized? This question is relevant because in the default cache we won't be able to offer that functionality instead we'd advise you to abstract the logic in some hook that listens to your websocket and appends the data to its own localstate.

In our normalized cache, we've been thinking about this functionality. There's even an open pr for it currently.

**Comment by mwangaben** [comment]
I am currently using default but if the normalized one will offer what i need i will go for it . I have also seen the PR please merge it it will be very useful
